lalrpop/src/lr1/interpreter.rs

```diff
@@ -0,0 +1,273 @@
+        Ok(())   
+    }
+
+    fn write_types(&mut self) -> io::Result<()> {
+        rust!(self.out, "struct ReducedProduction {{");
+        rust!(self.out, "nonterminal: u32,");
+        rust!(self.out, "symbol_count: u32,");
+        rust!(self.out, "}}");
+        Ok(())   
+    }
+
+    fn write_uses(&mut self) -> io::Result<()> {
+        try!(self.out.write_uses("super::", &self.grammar));
+        if self.grammar.intern_token.is_none() {
+            rust!(self.out, "use super::{}ToTriple;", self.prefix);
+        }
+        Ok(())
+    }
+
+    fn write_start_fn(&mut self) -> io::Result<()> {
+        let error_type = self.types.error_type();
+        let parse_error_type = self.parse_error_type();
+
+        let (type_parameters, parameters);
+
+        if self.grammar.intern_token.is_some() {
+            // if we are generating the tokenizer, we just need the
+            // input, and that has already been added as one of the
+            // user parameters
+            type_parameters = vec![];
+            parameters = vec![];
+        } else {
+            // otherwise, we need an iterator of type `TOKENS`
+            let mut user_type_parameters = String::new();
+            for type_parameter in &self.grammar.type_parameters {
+                user_type_parameters.push_str(&format!("{}, ", type_parameter));
+            }
+            type_parameters = vec![
+                format!("{}TOKEN: {}ToTriple<{}Error={}>",
+                        self.prefix, self.prefix, user_type_parameters, error_type),
+                format!("{}TOKENS: IntoIterator<Item={}TOKEN>", self.prefix, self.prefix)];
+            parameters = vec![format!("{}tokens: {}TOKENS", self.prefix, self.prefix)];
+        }
+
+        try!(self.out.write_pub_fn_header(
+            self.grammar,
+            format!("parse_{}", self.user_start_symbol),
+            type_parameters,
+            parameters,
+            format!("Result<{}, {}>",
+                    self.types.nonterminal_type(self.start_symbol),
+                    parse_error_type),
+            vec![]));
+        rust!(self.out, "{{");
+
+        if self.grammar.intern_token.is_some() {
+            // if we are generating the tokenizer, create a matcher as our input iterator
+            rust!(self.out, "let mut {}tokens = super::{}intern_token::{}Matcher::new(input);",
+                  self.prefix, self.prefix, self.prefix);
+        } else {
+            // otherwise, convert one from the `IntoIterator`
+            // supplied, using the `ToTriple` trait which inserts
+            // errors/locations etc if none are given
+            rust!(self.out, "let {}tokens = {}tokens.into_iter();", self.prefix, self.prefix);
+            rust!(self.out, "let mut {}tokens = {}tokens.map(|t| {}ToTriple::to_triple(t));",
+                  self.prefix, self.prefix, self.prefix);
+        }
```

Comment:
I wonder if we can/should abstract this out; probably yes, but also probably not worth it just now.